### PR TITLE
Unittests 

### PR DESCRIPTION
**Pull request by DangaRanga**
Unittests for chart, favorite, chart_datapoint, user, systemd and db.py

### CODE REVIEW

File: setup/_pattoo/systemd.py [hunk 1]
```diff
@@ -275,25 +273,20 @@ def run_systemd():
...
-        shared.run_script('systemctl enable pattoo_ingesterd')
-        shared.run_script('systemctl start pattoo_apid')
-        shared.run_script('systemctl start pattoo_api_agentd')
-        shared.run_script('systemctl start pattoo_ingesterd')
+    # Say what we are doing
+    print('??: Enabling system daemons')
+    # Reloading daemons
+    shared.run_script('sudo systemctl daemon-reload')
+    # Enabling daemons
+    shared.run_script('sudo systemctl enable pattoo_apid')
```
Comment (palisadoes) [inline setup/_pattoo/systemd.py:281]: Fails if daemon is already enabled. You need to fix this.

File: setup/_pattoo/systemd.py [hunk 2]
```diff
@@ -275,25 +273,20 @@ def run_systemd():
...
+    shared.run_script('sudo systemctl enable pattoo_api_agentd')
```
Comment (palisadoes) [inline setup/_pattoo/systemd.py:282]: Fails if daemon is already enabled. You need to fix this.

File: setup/_pattoo/systemd.py [hunk 3]
```diff
@@ -275,25 +273,20 @@ def run_systemd():
...
+    shared.run_script('sudo systemctl enable pattoo_ingesterd')
```
Comment (palisadoes) [inline setup/_pattoo/systemd.py:283]: Fails if daemon is already enabled. You need to fix this.

File: .travis.yml [hunk 4]
```diff
@@ -26,7 +26,12 @@ before_install:
     - pip3 install -r pip_requirements.txt
     - tests/bin/unittest_setup.py
     - mysql -e 'CREATE DATABASE pattoo_unittest;'
-    - setup/pattoo_installation.py install all
+    - wget https://bootstrap.pypa.io/get-pip.py
+    - sudo python3.6 get-pip.py
```
Comment (palisadoes) [inline .travis.yml:30]: What's the purpose of this?

File: setup/_pattoo/packages.py [hunk 5]
```diff
@@ -19,21 +20,25 @@ def install_missing_pip3(package, pip_dir, verbose=True):
...
+    # Validate pip directory
+    if not os.path.isdir(pip_dir):
+        shared.log('Pip directory is invalid')
     # Installs to the directory specified as pip_dir if the user is not travis
     username = getpass.getuser()
     if username == 'root':
         shared.run_script(
-            'python3 -m pip install {0} -t {1}'.format(package, pip_dir),
+            'python3 -m pip install {0} -t {1} -U --force-reinstall'.format(
+                                                                package,
```
Comment (palisadoes) [inline setup/_pattoo/packages.py:31]: The indentation here seems wrong. Please fix.

File: tests/setup_/_pattoo/test_configure.py [hunk 6]
```diff
@@ -69,17 +70,38 @@ def test_user_exists(self):
...
-
-    def test_check_pattoo_client(self):
-        """Unittest to test the check_pattoo_client function."""
-        pass
-
     def test_pattoo_server_config(self):
         """Unittest to test the pattoo_server_config function."""
-        pass
+        # Initialize key variables
+        expected = set([
```
Comment (palisadoes) [inline tests/setup_/_pattoo/test_configure.py:76]: This is more easily done using a triple quoted string. set() is not necessary. Load the configuration file as a string too.

File: tests/setup_/_pattoo/test_systemd.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+            actual_contents = sorted(
+                [temp_api_agentd, temp_ingesterd, temp_apid])
+
+            # Ensure contents are equal
+            self.assertEqual(expected_contents, actual_contents)
+
+    def test__symlink_dir(self):
+        """Testing method or function named "_symlink_dir"."""
+        # Initialise key variables
+        etc_dir = '/etc/systemd/system/multi-user.target.wants'
```
Comment (palisadoes) [inline tests/setup_/_pattoo/test_systemd.py:113]: This location depends on the operating system. In CentOS it is '/usr/lib/systemd/' You need to test for this in ALL cases including installation. The OS will not always be Ubuntu

File: .travis.yml (hunk 4 above)
Comment (DangaRanga) [reply .travis.yml:30]: When I tried executing the installation script as sudo in travis, the build would fail due to pip being missing, so I had to install pip in travis' sudo environment to allow for the packages to install whenever the script was run as sudo

File: setup/_pattoo/packages.py [hunk 8]
```diff
@@ -27,9 +27,7 @@ def install_missing_pip3(package, pip_dir, verbose=False):
     username = getpass.getuser()
     if username == 'root':
         shared.run_script(
-            'python3 -m pip install {0} -t {1} -U --force-reinstall'.format(
-                                                                package,
-                                                                pip_dir),
+            'python3 -m pip install {0} -t {1} -U'.format(package, pip_dir),
```
Comment (palisadoes) [inline setup/_pattoo/packages.py:30]: Why did you remove `--force-reinstall` ?

File: setup/_pattoo/shared.py [hunk 9]
```diff
@@ -65,10 +65,12 @@ def run_script(cli_string, die=True, verbose=True):
             )
 
         # Log message
-        print('messages: {}'.format(messages))
+        if verbose is True:
+            print('messages: {}'.format(messages))
         if messages != []:
```
Comment (palisadoes) [inline setup/_pattoo/shared.py:70]: Use `if bool(messages) is False:`

File: setup/_pattoo/systemd.py [hunk 10]
```diff
@@ -263,7 +263,33 @@ def preflight(config_dir, etc_dir):
...
+    """
+    for daemon in daemons:
+        # Initialize key variables
+        symlink_path = os.path.join(etc_dir, daemon)
+        file_path = os.path.join(symlink_dir, daemon)
+
+        # Say what we are doing
+        print('Checking if the {}.service file is a symlink '.format(daemon))
+        link = os.path.islink('{0}.service'.format(symlink_path))
+        if link is False:
```
Comment (palisadoes) [inline setup/_pattoo/systemd.py:286]: use the `systemctl` command to create the symlink. That way you are certain it'll be operating system compatible.

File: tests/setup_/_pattoo/test_configure.py [hunk 11]
```diff
@@ -69,17 +70,29 @@ def test_user_exists(self):
...
-
-    def test_check_pattoo_client(self):
-        """Unittest to test the check_pattoo_client function."""
-        pass
-
     def test_pattoo_server_config(self):
         """Unittest to test the pattoo_server_config function."""
-        pass
+        # Initialize key variables
+        expected = '''\
```
Comment (palisadoes) [inline tests/setup_/_pattoo/test_configure.py:76]: You can use this
```

'''\
pattoo_api_agentd:
ip_bind_port: 20201
ip_listen_address: 0.0.0.0
pattoo_apid:
  ip_bind_port: 20202
ip_listen_address: 0.0.0.0
pattoo_db:
db_hostname: localhost
  db_max_overflow: 20
db_name: pattoo
  db_password: password
db_pool_size: 10
  db_username: pattoo
pattoo_ingesterd:
batch_size: 500
  graceful_timeout: 10
ingester_interval: 3600
'''

```

File: tests/setup_/_pattoo/test_packages.py [hunk 12]
```diff
@@ -63,22 +65,32 @@ def test_install_missing_pip3(self):
...
+            # At least one expected package
+            expected_package = 'PattooShared'
+            expected = True
+
+            # Create temporary directory
+            with tempfile.TemporaryDirectory() as temp_dir:
+                result = install(ROOT_DIR, temp_dir)
+
+                # Get raw packages in requirements format
+                packages = subprocess.check_output(
```
Comment (palisadoes) [inline tests/setup_/_pattoo/test_packages.py:86]: Why did you use this instead of the runscript function?

File: setup/_pattoo/systemd.py [hunk 13]
```diff
@@ -275,25 +299,27 @@ def run_systemd():
...
+    # Say what we are doing
+    print('Checking if daemons are already running')
+
+    # Loop through daemon list
+    for daemon in daemons:
+        command = '''
+systemctl is-active {} daemon --quiet service-name'''.format(daemon)
+
+        # Check status code of daemon
+        status = os.system(command)
```
Comment (palisadoes) [inline setup/_pattoo/systemd.py:311]: Avoid using os.system as a rule, modify runscript to figure out the exit code. The advantage is that runscript will also give you access to the error messages that os.system won't. Make runscript return an exit code for example, if it doesn't already. Consider a failsafe parameter for runscript so that it doesn't die under certain conditions, if it doesn't already.

File: setup/_pattoo/packages.py (hunk 8 above)
Comment (DangaRanga) [reply setup/_pattoo/packages.py:30]: My bad, I confused  how --upgrade worked and thought it reinstalled the packages without the --force-reinstall tag
I'm going to add it back in the next commit

File: tests/setup_/_pattoo/test_packages.py (hunk 12 above)
Comment (DangaRanga) [reply tests/setup_/_pattoo/test_packages.py:86]: my bad, I didn't realise that the runscript function also stored the output at first. I just made a modification that uses it instead of check_output

File: setup/_pattoo/systemd.py (hunk 10 above)
Comment (DangaRanga) [reply setup/_pattoo/systemd.py:285]: For unittesting I might have to use os.symlink() function to set the symlinks, and test if they're present, since the systemctl command needs sudo privileges, is this okay?
